### PR TITLE
Check for illegal characters when formatting names

### PR DESCRIPTION
**Pull request by parsonsmatt**
This PR checks the type name for illegal characters before generating code.

### CODE REVIEW

**Comment by thomasjm** [comment]
Thanks! `checkIllegalName` seems not to be used anywhere though :P

Also, could you move the function to `Util.hs` or perhaps `Formatting.hs`?

FWIW, if something like this has a tendency to complicate the core code in `TH.hs`, it could also make sense to do it as part of the formatting step, e.g. the `formatTSDeclaration` function. At the formatting stage, the data types and logic are simpler and the TS names for things are more apparent. Might need an API change to allow `formatTSDeclaration` to error out in that case, although a pure exception doesn't seem out of the question--maybe a new function `formatTSDeclarationEither`. This all depends on whether it can be done cleanly in `TH.hs` of course.

**Comment by parsonsmatt** [comment]
Lol oops!

I put the function in the `Util` module as requested, and refactored it so it could be tested a bit more easily.

🤔 

Right now, there's no option to customize the Haskell type name. So its *definitely* going to be a problem if we have any illegal characters there, and we can easily report the error as soon as we have the `Name` of the type we're deriving for.

I've modified the PR to also check datatype constructors.

---

And at this point I recognize that I've got a problem - the *formatting options* allow you to modify the interface and type names, which happens well after `TemplateHaskell` is done. That's a problem! We can't actually check at compile-time that the generated TS will be "correct" because we might *change* the typescript that actually gets written - `getTypeScriptDeclarations` for a given type isn't guaranteed to return the same typescript code that is actually printed by the app.

With this patch, anyone that is currently doing `filter ('\'' ==)` as a name modifier would get a compile-time error that doesn't really apply to them.

Oof.

I'd propose to modify the `FormattingOptions` and move those name formatters into the `ExtraTypeScriptOptions`. That gives you better consistency over the generated names, as well as allowing you to know at compile-time if the names are invalid.

**Comment by thomasjm** [comment]
Oh right, that's unfortunate. I don't love the idea of a breaking change to the name formatters API, although in hindsight it seems like the `ExtraTypeScriptOptions` might have been a better place for them. FWIW here's the PR that added the formatters: https://github.com/codedownio/aeson-typescript/pull/5.

As an alternative hack, what about just replacing all invalid characters with underscores at formatting time? :)

**Comment by parsonsmatt** [comment]
> As an alternative hack, what about just replacing all invalid characters with underscores at formatting time? :)

I think our linter complains about trailing `_` 😅 

A default name formatter that throws an error on an invalid character + tells you how to fix it may be the least invasive solution - we can change the formatter to be something like:

```haskell
defaultNameFormatter str = do
  case NonEmpty.nonEmpty str of
    Nothing -> error "Name cannot be empty"
    Just nameChars -> case checkIllegalNameChars nameChars of
      Just badChars -> error $ concat ["The name ", str, " contains illegal characters: ", toList badChars, "\nConsider setting a default name formatter that replaces these characters." ]
      Nothing -> str
``` 

Obviously would want to trace the "how did i get here" so users can backtrack to exactly what failed + which name formatter needs to be invoked.

I do think the best choice is moving the formatting options to the `TH` side. Then you'd get a compile-time error, with a note to modify the name transformation field *or* the type itself. Given two steps, it wouldn't be a hard migration for end users:

1. Next major version: Release the new name transformation fields on `ExtraTypeScriptOptions` and deprecate the formatter fields w/ good migration messages
2. Major version after that: Delete the formatters.

**Comment by thomasjm** [comment]
Setting the default name formatter to error out as you describe sounds like a good first step. Maybe that would be sufficient to close #34 ?

If you get an informative error message at the formatting stage, would you still feel strongly about moving those options to the TH side?

**Comment by parsonsmatt** [comment]
Yeah, I generally want problems reported ASAP. 

At least for our codebase, the derivation happens *well* before code generation, which is a totally separate step. I wouldn't want to introduce a runtime dependency on `tsc` for our test suite just to ensure that our generated code is well-formed.

**Comment by thomasjm** [comment]
> I wouldn't want to introduce a runtime dependency on `tsc`

Huh? There's no `tsc` involved in the idea of erroring out at the formatting stage by using `checkIllegalNameChars`.

Just thinking through the implications of moving the formatting options -- it could harm at least one workflow I can imagine. What if you have lots of primed types like `Foo'` and `Bar'`, and you're accustomed to bulk-formatting them by calling `formatTSDeclarations'` with an `interfaceNameModifier` that does `T.replace "'" "Prime"`? If we move the options, then such a user would have to individually apply formatting options to all their types at the `deriveTypeScript` call. (Assuming most people  derive instances individually and format in bulk, like I do.) I suppose if we go through the deprecation process then people will be free to complain at that point.

**Comment by parsonsmatt** [comment]
Yeah, that would require a migration of code - but it shouldn't *have* to be a hard one. At work we consolidate our default options, so we can easily control what that default is by altering a single value in a single module. Factoring out the duplication in argument passing wouldn't be a huge problem.

**Comment by thomasjm** [comment]
All right, what do you think of the following?

1) In this PR, do the erroring `defaultNameFormatter`
2) Make a subsequent PR moving the formatting functions to the `ExtraTypeScriptOptions`, and maybe ask for comments from @Reisen

**Comment by parsonsmatt** [comment]
Sounds good! I'll make the relevant changes.

**Comment by parsonsmatt** [comment]
OK, should be ready for review :)

File: src/Data/Aeson/TypeScript/LegalName.hs [hunk 1]
```diff
@@ -0,0 +1,58 @@
+-- | This module defines functions which are useful for determing if
```
Comment (thomasjm) [inline src/Data/Aeson/TypeScript/LegalName.hs:1]: Typo: determining

File: src/Data/Aeson/TypeScript/LegalName.hs [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+-- a given name is a legal JavaScript name accord to <https://stackoverflow.com/questions/1661197/what-characters-are-valid-for-javascript-variable-names this post>.
```
Comment (thomasjm) [inline src/Data/Aeson/TypeScript/LegalName.hs:2]: Typo: according

File: src/Data/Aeson/TypeScript/LegalName.hs [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+import Language.Haskell.TH
+import Data.List.NonEmpty (NonEmpty (..))
+import qualified Data.List.NonEmpty as NonEmpty
+import Data.Char
+import Data.Foldable
+
+
+-- | This reports a compile-time error if the given 'Name' contains
+-- characters that are not allowed in a JavaScript name.
+checkIllegalName :: Name -> Q ()
```
Comment (thomasjm) [inline src/Data/Aeson/TypeScript/LegalName.hs:15]: The TH functions probably don't belong in this PR. Do we want to export this whole module?

File: src/Data/Aeson/TypeScript/Types.hs [hunk 4]
```diff
@@ -131,12 +133,30 @@ data SumTypeFormat =
...
+  , interfaceNameModifier = defaultNameFormatter
+  , typeNameModifier = defaultNameFormatter
   , exportMode = ExportNone
   , typeAlternativesFormat = TypeAlias
   }
 
+-- | The 'defaultNameFormatter' in the 'FormattingOptions' checks to see if
+-- the name is a legal TypeScript name. If it is not, then it throws
+-- a runtime error.
+defaultNameFormatter :: String -> String
```
Comment (thomasjm) [inline src/Data/Aeson/TypeScript/Types.hs:145]: This should probably be exported next to `defaultFormattingOptions` in `TH.hs`.

File: test/Util.hs [hunk 5]
```diff
@@ -15,6 +15,7 @@ import System.Exit
 import System.FilePath
 import System.IO.Temp
 import System.Process
+import Test.Hspec
```
Comment (thomasjm) [inline test/Util.hs:18]: Why is this needed? Seems redundant for me on Stackage LTS 20.1

**Review by thomasjm — CHANGES_REQUESTED** [review]
Looks great! Just a few comments.

**Comment by thomasjm** [comment]
Sorry, this fell off my radar -- merging now!

**Comment by thomasjm** [comment]
Automatic merge failed, manually rebased and merged in ba5d998

**Pull request closed without merging**